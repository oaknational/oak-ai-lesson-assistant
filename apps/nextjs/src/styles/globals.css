@tailwind base;
@tailwind components;
@tailwind utilities;

.radix-themes {
  --default-font-family: Lexend, ui-sans-serif, system-ui, -apple-system,
    BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans",
    sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol",
    "Noto Color Emoji";
}
[data-state="active"][role="tab"] {
  background-color: #222222;
  border-radius: 4px;
  color: white; /* Change text color if desired */
}

@layer utilities {
  .no-scrollbar {
    -ms-overflow-style: none; /* IE and Edge */
    scrollbar-width: none; /* Firefox */
  }
}
