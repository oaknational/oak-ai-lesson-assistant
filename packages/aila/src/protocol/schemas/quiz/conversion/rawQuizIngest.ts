import { aiLogger } from "@oakai/logger";

import invariant from "tiny-invariant";

import type { QuizV2, QuizV2Question } from "../quizV2";
import type {
  HasuraQuiz,
  StemImageObject,
  StemObject,
  StemTextObject,
} from "../rawQuiz";

const log = aiLogger("aila:quiz");

/**
 * Check if an item is a text item
 */
function isTextItem(item: StemObject): item is StemTextObject {
  return item.type === "text";
}

/**
 * Check if an item is an image item
 */
function isImageItem(item: StemObject): item is StemImageObject {
  return item.type === "image";
}

/**
 * Extract markdown content from Oak's content items array, inlining images
 * Returns both the markdown content and attribution metadata
 */
function extractMarkdownFromContent(
  contentItems: Array<StemObject | undefined>,
): {
  markdown: string;
  attributions: Array<{ imageUrl: string; attribution: string }>;
} {
  const attributions: Array<{ imageUrl: string; attribution: string }> = [];

  const markdownParts = contentItems
    .filter((item): item is StemObject => item !== undefined)
    .map((item) => {
      if (isTextItem(item)) {
        return item.text || "";
      } else if (isImageItem(item)) {
        const imageUrl = item.image_object.secure_url;

        // Extract alt text from context if available
        const altText = item.image_object.context?.custom?.alt ?? "";

        // Extract attribution if available
        if (
          item.image_object.metadata &&
          typeof item.image_object.metadata === "object" &&
          !Array.isArray(item.image_object.metadata)
        ) {
          const attribution = item.image_object.metadata.attribution;
          if (attribution) {
            attributions.push({ imageUrl, attribution });
          }
        }

        // Return markdown image syntax with alt text
        return `![${altText}](${imageUrl})`;
      }
      return "";
    });

  return {
    markdown: markdownParts.join(" ").trim(),
    attributions,
  };
}

/**
 * Convert Hasura quiz format to Quiz V2 format
 */
export function convertHasuraQuizToV2(hasuraQuiz: HasuraQuiz): QuizV2 {
  log.info("convertHasuraQuizToV2 input:", { hasuraQuiz });

  if (!hasuraQuiz || !Array.isArray(hasuraQuiz)) {
    log.info("Hasura quiz is not an array, returning empty quiz");
    return {
      version: "v2",
      questions: [],
      imageAttributions: [],
    };
  }

  // Collect all image attributions from all questions
  const allImageAttributions: Array<{ imageUrl: string; attribution: string }> =
    [];

  const questions = hasuraQuiz
    .filter(
      (hasuraQuestion) => hasuraQuestion.questionType !== "explanatory-text",
    )
    .map((hasuraQuestion): QuizV2Question => {
      log.info("Processing hasura question:", { hasuraQuestion });

      // Early return for explanatory-text (should be filtered out already)
      if (hasuraQuestion.questionType === "explanatory-text") {
        throw new Error("Explanatory text questions should be filtered out");
      }
      // Extract question stem as markdown with inlined images
      const { markdown: questionStem, attributions } =
        extractMarkdownFromContent(hasuraQuestion.questionStem);

      const hint = hasuraQuestion.hint ?? null;

      // Handle different question types based on Oak's schema
      switch (hasuraQuestion.questionType) {
        case "multiple-choice": {
          const mcAnswers = hasuraQuestion.answers?.["multiple-choice"] ?? [];

          const correctAnswerResults = mcAnswers
            .filter((answer) => answer.answer_is_correct)
            .map((answer) => {
              invariant(
                answer.answer,
                "Multiple choice answer missing 'answer' field",
              );
              return extractMarkdownFromContent(answer.answer);
            });
          const correctAnswers = correctAnswerResults.map(
            (result) => result.markdown,
          );

          const distractorResults = mcAnswers
            .filter((answer) => !answer.answer_is_correct)
            .map((answer) => {
              invariant(
                answer.answer,
                "Multiple choice answer missing 'answer' field",
              );
              return extractMarkdownFromContent(answer.answer);
            });
          const distractors = distractorResults.map(
            (result) => result.markdown,
          );

          // Collect all attributions from question and answers
          allImageAttributions.push(
            ...attributions,
            ...correctAnswerResults.flatMap((result) => result.attributions),
            ...distractorResults.flatMap((result) => result.attributions),
          );

          return {
            questionType: "multiple-choice" as const,
            question: questionStem,
            answers: correctAnswers,
            distractors,
            hint,
          };
        }

        case "short-answer": {
          const saAnswers = hasuraQuestion.answers?.["short-answer"] ?? [];
          const answerResults = saAnswers.map((answer) => {
            invariant(answer.answer, "Short answer missing 'answer' field");
            return extractMarkdownFromContent(answer.answer);
          });
          const answers = answerResults.map((result) => result.markdown);

          allImageAttributions.push(
            ...attributions,
            ...answerResults.flatMap((result) => result.attributions),
          );

          return {
            questionType: "short-answer" as const,
            question: questionStem,
            answers,
            hint,
          };
        }

        case "match": {
          const matchAnswers = hasuraQuestion.answers?.match ?? [];
          const pairs = matchAnswers.map((matchItem) => {
            const leftResult = extractMarkdownFromContent(
              matchItem.match_option ?? [],
            );
            const rightResult = extractMarkdownFromContent(
              matchItem.correct_choice || [],
            );

            allImageAttributions.push(
              ...leftResult.attributions,
              ...rightResult.attributions,
            );

            return {
              left: leftResult.markdown,
              right: rightResult.markdown,
            };
          });

          allImageAttributions.push(...attributions);

          return {
            questionType: "match" as const,
            question: questionStem,
            pairs,
            hint,
          };
        }

        case "order": {
          const orderAnswers = hasuraQuestion.answers?.order ?? [];
          const itemResults = orderAnswers.map((orderItem) =>
            extractMarkdownFromContent(orderItem.answer || []),
          );
          const items = itemResults.map((result) => result.markdown);

          allImageAttributions.push(
            ...attributions,
            ...itemResults.flatMap((result) => result.attributions),
          );

          return {
            questionType: "order" as const,
            question: questionStem,
            items,
            hint,
          };
        }

        default: {
          const _exhaustiveCheck: never = hasuraQuestion.questionType;
          throw new Error(
            `Unknown question type: ${_exhaustiveCheck as string}`,
          );
        }
      }
    });

  return {
    version: "v2",
    questions,
    imageAttributions: allImageAttributions,
  };
}
