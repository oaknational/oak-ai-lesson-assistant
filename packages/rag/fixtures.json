{
  "inventions of the industrial revolution": {
    "embedding": [
      0.024054011, 0.0109390505, -0.043416213, -0.038962398, -0.1653692,
      -0.04736005, -0.14959385, -0.099207915, 0.0029982517, 0.082616605,
      0.0067402227, 0.058205605, 0.08098467, 0.041988272, 0.032060683,
      0.057797622, -0.06983313, -0.013480445, -0.038554415, -0.041920274,
      -0.009349615, 0.00036017268, -0.026195923, -0.08853236, -0.047734037,
      -0.03347162, 0.011109043, -0.021130132, -0.065855294, 0.14700995,
      0.029595783, -0.0437902, -0.0032043683, -0.027878853, 0.08084867,
      0.026127925, 0.048312012, 0.057933617, 0.10471569, -0.09628404,
      0.042362258, -0.026586907, 0.020807143, 0.014746893, -0.03699048,
      0.071057074, -0.13735434, -0.020314164, -0.11511926, 0.117227174,
      -0.0049297973, -0.12130701, 0.001131516, 0.054125775, -0.04640809,
      -0.058409598, 0.04348421, 0.13823831, 0.010038087, -0.13477045,
      0.0069314647, 0.035698533, -0.16400926, 0.052187853, -0.013276454,
      -0.07636086, -0.071125075, -0.015758352, 0.110699445, -0.003935338,
      0.002284281, 0.06639927, 0.025379956, 0.041478295, 0.08322857,
      -0.070921086, 0.019855183, 0.04695207, 0.051609877, 0.022864059,
      0.091932215, 0.01893722, 0.00899263, -0.06361138, 0.055213727,
      0.028541826, 0.04181828, -0.020246167, 0.121511005, 0.054635752,
      0.009001129, 0.09302017, 0.0534458, -0.0007793117, 0.010361074,
      0.030700738, 0.07506891, -0.059939533, -0.013344451, 0.035902523,
      -0.023782022, 0.0315507, -0.013573942, -0.077652805, 0.058001615,
      0.012256496, -0.019532196, 0.08057669, -0.10131583, 0.09839195,
      0.023527032, -0.025396956, 0.04086632, -0.017373286, -0.060177524,
      -0.038282424, 0.020620152, -0.105055675, 0.010157082, 0.01893722,
      0.06639927, 0.0066424767, 0.07262101, -0.013896928, -0.103015766,
      -0.010641562, 0.0631354, -0.11348733, -0.03811243, -0.039336383,
      -0.03211168, 0.022898057, 0.01602184, 0.051643874, 0.07683684,
      -0.07377697, 0.089144334, -0.0021153504, -0.063577384, 0.0631354,
      -0.12096702, -0.0018295498, -0.016327828, -0.06850718, -0.011007047,
      0.0061834957, -0.15530561, -0.005864759, -0.06031352, 0.059327558,
      -0.110223465, -0.014228415, -0.083772555, 0.038214426, -0.01118554,
      0.014959385, 0.0153758675, 0.037330464, -0.016089838, -0.08098467,
      0.051167894, 0.037500456, -0.06765722, 0.06939115, 0.116547205,
      -0.038248427, 0.07214503, 0.04253225, -0.029527785, 0.0340496, 0.07568089,
      0.011797515, 0.0050147935, -0.16754511, 0.08248061, -0.059667546,
      -0.0048660496, -0.0135229435, 0.012477486, -0.06446135, -0.066161275,
      0.031074721, 0.033794608, 0.034508582, 0.067895204, 0.0047385553,
      -0.0063067405, 0.036718488, 0.04277024, -0.07874076, -0.023068052,
      -0.10233579, -0.014984883, 0.1650972, -0.037670452, -0.021504115,
      -0.003478482, -0.027249878, -0.097575985, -0.0046620583, -0.076292865,
      0.10716359, 0.06541331, -0.052833825, 0.050929904, 0.010922051,
      -0.024105009, 0.084520526, 0.03818043, -0.060041532, 0.018206252,
      -0.024342999, -0.088668354, -0.063645385, -0.048006024, 0.040526334,
      0.11974307, 0.019804185, -0.08975631, -0.02798085, 0.0085293995,
      0.027606864, 0.028677821, 0.0046620583, -0.051405884, -0.027079886,
      -0.10845554, 0.06361138, 0.049773954, 0.005053042, 0.035664532,
      -0.15666555, -0.1264068, 0.052425843, -0.078400776, 0.055927698,
      -0.043892194, 0.0052400343, -0.014313411, -0.022660067, 0.039778363,
      -0.0021079134, -0.010871053, -0.08547249, 0.011933509, -0.015290871,
      -0.009035128, 0.0947201, 0.05470375, -0.027997848, 0.058851577,
      -0.0028473828, 0.010488569, -0.08554048, -0.07214503, 0.04103631
    ]
  },
  "the end of roman britain": {
    "embedding": [
      -0.06042437, -0.061291862, -0.015873492, 0.050047792, -0.06189244,
      -0.021437138, -0.0015108178, 0.05979043, -0.082412034, 0.036034413,
      -0.060891483, 0.08087724, -0.087950654, -0.04317456, 0.026208362,
      -0.09776002, 0.0021218178, -0.000719437, 0.02659206, -0.08161127,
      -0.043141197, 0.020069165, -0.03653489, -0.089618914, -0.09896117,
      0.023038667, -0.03146338, -0.06496204, -0.029328007, 0.09615849,
      0.021303678, -0.02472361, 0.17229787, 0.054351903, -0.016107049,
      0.13799845, 0.1247191, 0.10263134, -0.036868542, -0.04751204, 0.026942395,
      -0.0071568345, -0.015523157, 0.04494292, 0.020669738, -0.017933793,
      0.018767923, -0.02742619, -0.059456777, -0.01778365, -0.04284091,
      0.0012793468, -0.032848034, 0.04784569, -0.048512995, -0.027759843,
      0.06906595, 0.036568254, -0.10630151, 0.008316275, -0.047311846,
      0.074871495, 0.11584396, 0.02499053, -0.091821015, -0.0139466515,
      0.08227857, -0.014146843, -0.07146825, -0.065228954, 0.033515338,
      0.019118257, 0.0041831615, -0.06185907, -0.12852274, -0.02298862,
      0.05848919, 0.03993814, -0.13312712, -0.008758364, -0.022421412,
      -0.0076948483, 0.03011209, -0.018033888, 0.036267966, -0.007699019,
      -0.054952476, -0.069132686, 0.04681137, -0.04861309, 0.05959024,
      -0.0468781, -0.019051526, -0.20286039, 0.015398038, -0.07140152,
      0.035233647, 0.016332263, -0.07233574, -0.06479521, -0.042807546,
      -0.045076378, 0.024456687, 0.05361787, 0.048846647, -0.05008116,
      0.00113963, -0.013963334, 0.022171173, 0.21180226, 0.12024816, 0.06696395,
      0.007653142, 0.038002957, -0.08354645, 0.0017495875, 0.028977672,
      -0.057821885, -0.036634985, 0.04244053, -0.024973849, -0.043341387,
      0.10863708, -0.0014180208, -0.044842824, 0.072869584, -0.03610114,
      0.039204106, -0.05395152, 0.005613694, -0.10016232, 0.021303678,
      0.05144913, 0.19164968, 0.056287084, 0.0042749154, -0.039404295,
      0.019034844, 0.09242159, -0.024690244, -0.0076906774, 0.06289339,
      0.0012845601, 0.032464337, 0.003634721, 0.03570076, 0.043374754,
      -0.028744116, -0.009675907, -0.05888957, 0.0026024852, 0.083880104,
      -0.0044459123, -0.035233647, -0.037168827, -0.0336488, 0.00944235,
      -0.033415243, -0.038303245, 0.04063881, -0.063160315, 0.04554349,
      -0.017057955, 0.009008603, -0.015681641, -0.07440439, 0.12445218,
      -0.03820315, -0.026425235, 0.008883484, 0.11397551, -0.06516223,
      -0.054185078, -0.044208884, 0.07867513, 0.033765577, -0.09569138,
      -0.0445759, -0.022171173, 0.09422331, 0.020002434, -0.017950475,
      -0.06215936, -0.05568651, 0.045043014, -0.008337128, -0.016198803,
      0.0074279266, -0.0558867, -0.08574855, -0.07680668, -0.12191642,
      0.12084874, 0.04514311, -0.076339565, -0.014588932, 0.073203236,
      0.08181146, -0.0468781, -0.027125904, 0.13933305, 0.054285172,
      -0.04327466, -0.08488106, 0.08367991, -0.03750248, 0.051849514,
      0.106701896, 0.15334643, -0.015131116, -0.0036242944, 0.04327466,
      -0.036768444, -0.0026921541, -0.087750465, 0.07240248, -0.034833264,
      0.012178296, -0.015556523, 0.021820838, -0.03306491, -0.044409074,
      -0.083146065, 0.04187332, 0.014680686, -0.052883834, 0.0075447047,
      0.08621567, 0.07827475, 0.053217486, -0.049547315, -0.00712764,
      -0.11771241, -0.0070400564, 0.07146825, -0.0028402123, -0.106902085,
      -0.0017214356, 0.07527188, 0.043341387, -0.025924757, 0.04631089,
      -0.03171362, -0.10756939, 0.05425181, 0.0051424108, -0.07580572,
      0.037402384, -0.05888957, -0.003622209, 0.1297906, 0.09408985,
      0.0053551137, 0.10550075, -0.080743775, 0.005496916
    ]
  }
}
