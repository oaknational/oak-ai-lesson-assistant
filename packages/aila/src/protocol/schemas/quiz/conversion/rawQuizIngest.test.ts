import { describe, expect, it } from "@jest/globals";

import { rawQuizFixture } from "../fixtures/rawQuizFixture";
import type { RawQuiz } from "../rawQuiz";
import { convertRawQuizToV2 } from "./rawQuizIngest";

describe("convertRawQuizToV2", () => {
  it("should convert raw quiz to V2 format", () => {
    const result = convertRawQuizToV2(rawQuizFixture);
    expect(result).toMatchSnapshot();
  });

  it("should handle empty quiz", () => {
    const result = convertRawQuizToV2([]);
    expect(result).toEqual({
      version: "v2",
      questions: [],
      imageAttributions: [],
    });
  });

  it("should handle null quiz", () => {
    const result = convertRawQuizToV2(null);
    expect(result).toEqual({
      version: "v2",
      questions: [],
      imageAttributions: [],
    });
  });

  it("should filter out explanatory-text questions", () => {
    const explanatoryOnlyQuiz: RawQuiz = [
      {
        questionId: 1,
        questionUid: "test-uid-1",
        questionType: "explanatory-text",
        questionStem: [{ text: "Just text", type: "text" }],
        answers: { "explanatory-text": null },
        feedback: "",
        hint: "",
        active: true,
      },
    ];
    const result = convertRawQuizToV2(explanatoryOnlyQuiz);
    expect(result.questions).toHaveLength(0);
  });

  it("should extract image attributions correctly", () => {
    const quizWithImages: RawQuiz = [
      {
        questionId: 1,
        questionUid: "test-uid-1",
        questionType: "multiple-choice",
        questionStem: [
          {
            image_object: {
              secure_url: "https://example.com/image1.jpg",
              metadata: {
                attribution: "Photo by Photographer",
              },
            },
            type: "image",
          },
        ],
        answers: {
          "multiple-choice": [
            {
              answer: [{ text: "A", type: "text" }],
              answer_is_correct: true,
            },
          ],
        },
        feedback: "",
        hint: "",
        active: true,
      },
    ];
    const result = convertRawQuizToV2(quizWithImages);
    expect(result.questions[0]?.question).toContain(
      "![](https://example.com/image1.jpg)",
    );
    expect(result.imageAttributions).toEqual([
      {
        imageUrl: "https://example.com/image1.jpg",
        attribution: "Photo by Photographer",
      },
    ]);
  });

  it("should convert images to markdown syntax", () => {
    const quizWithImages: RawQuiz = [
      {
        questionId: 1,
        questionUid: "test-uid-1",
        questionType: "multiple-choice",
        questionStem: [
          { text: "Look at", type: "text" },
          {
            image_object: {
              secure_url: "https://example.com/image.jpg",
              metadata: {},
            },
            type: "image",
          },
          { text: "What is it?", type: "text" },
        ],
        answers: {
          "multiple-choice": [
            {
              answer: [{ text: "A", type: "text" }],
              answer_is_correct: true,
            },
          ],
        },
        feedback: "",
        hint: "",
        active: true,
      },
    ];
    const result = convertRawQuizToV2(quizWithImages);
    expect(result.questions[0]?.question).toBe(
      "Look at ![](https://example.com/image.jpg) What is it?",
    );
  });

  it("should extract alt text from images when available", () => {
    const quizWithAltText: RawQuiz = [
      {
        questionId: 1,
        questionUid: "test-uid-1",
        questionType: "multiple-choice",
        questionStem: [
          {
            image_object: {
              secure_url: "https://example.com/dog.jpg",
              metadata: {},
              context: {
                custom: {
                  alt: "A golden retriever sitting in a park",
                },
              },
            },
            type: "image",
          },
        ],
        answers: {
          "multiple-choice": [
            {
              answer: [{ text: "Dog", type: "text" }],
              answer_is_correct: true,
            },
          ],
        },
        feedback: "",
        hint: "",
        active: true,
      },
    ];
    const result = convertRawQuizToV2(quizWithAltText);
    expect(result.questions[0]?.question).toBe(
      "![A golden retriever sitting in a park](https://example.com/dog.jpg)",
    );
  });

  it("should throw error for unknown question type", () => {
    const unknownTypeQuiz = [
      {
        questionId: 1,
        questionUid: "test-uid-1",
        questionType: "unknown-type" as unknown as "multiple-choice",
        questionStem: [{ text: "Question", type: "text" }],
        answers: {},
        feedback: "",
        hint: "",
        active: true,
      },
    ] as RawQuiz;
    expect(() => convertRawQuizToV2(unknownTypeQuiz)).toThrow(
      "Unknown question type: unknown-type",
    );
  });
});
