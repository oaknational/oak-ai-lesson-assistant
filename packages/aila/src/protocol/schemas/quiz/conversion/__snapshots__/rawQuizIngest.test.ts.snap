// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`convertHasuraQuizToV3 should convert Hasura quiz to V3 format 1`] = `
{
  "imageMetadata": [
    {
      "attribution": "Photo by John Doe on Unsplash",
      "height": 768,
      "imageUrl": "https://example.com/dog.jpg",
      "width": 1024,
    },
    {
      "attribution": "Image Â© Example Inc",
      "height": 480,
      "imageUrl": "https://example.com/cat.jpg",
      "width": 640,
    },
  ],
  "questions": [
    {
      "answers": [
        "Paris",
      ],
      "distractors": [
        "London",
        "Berlin",
      ],
      "hint": "Think about the Eiffel Tower",
      "question": "What is the capital of France?",
      "questionType": "multiple-choice",
    },
    {
      "answers": [
        "red",
        "blue",
        "yellow",
      ],
      "hint": "These colors cannot be created by mixing other colors",
      "question": "Name a primary color",
      "questionType": "short-answer",
    },
    {
      "answers": [
        "Dog",
      ],
      "distractors": [
        "Cat ![](https://example.com/cat.jpg)",
      ],
      "hint": "It barks",
      "question": "Which animal is shown in the image? ![A friendly golden retriever sitting in a park](https://example.com/dog.jpg)",
      "questionType": "multiple-choice",
    },
    {
      "hint": "Think about European geography",
      "pairs": [
        {
          "left": "France",
          "right": "Paris",
        },
        {
          "left": "Germany",
          "right": "Berlin",
        },
      ],
      "question": "Match the countries to their capitals",
      "questionType": "match",
    },
    {
      "hint": "Start with the smallest",
      "items": [
        "1",
        "3",
        "5",
      ],
      "question": "Put these numbers in ascending order",
      "questionType": "order",
    },
  ],
  "version": "v3",
}
`;
